[6.0] Fixes the missing modified start and end date filter to articled API endpoint

laoneo opened the pull request:
### Summary of Changes
Adds two new filter options to the articles list endpoint. The new `modified_start` and `modified_end` filters do allow to fetch articles which are modified within a date range. This is a common use case in API's, as it allows consumers to fetch only the delta which has changed since the last sync and not always the full list of articles. The following example fetches only articles which are edited within February 2025:

`GET /api/index.php/v1/content/articles?filter[modified_start]=2025-02-01&filter[modified_start]=2025-02-28`

### Testing Instructions
- Create an article with the modified date before February 2025
- Create an article with the modified date after February 2025

Use tools like postman to execute the following curl command against the joomla installation with the patch:

` curl -g "https://{web site}/api/index.php/v1/content/articles?filter[modified_start]=2025-02-01&filter[modified_start]=2025-02-28" -H "Authorization: Bearer {your token}"`

### Actual result BEFORE applying this Pull Request
All articles are returned.

### Expected result AFTER applying this Pull Request
Only the article which was modified after the first February is returned.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

laoneo committed:
Add modified start and end date filter to articled API endpoint

brianteeman commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -492,6 +492,20 @@ protected function getListQuery()
         }
 
+        // Filter by date after modified date.
+        $modified = $this->getState('filter.modified_start');
+
+        if (!empty($modified)) {
+            $query->where($db->quoteName('a.modified') . ' >= ' . $db->quote($modified));
+        }
+
+        // Filter by date before modified date.
+        $modified = $this->getState('filter.modified_end');
+
+        if (!empty($modified)) {
+            $query->where($db->quoteName('a.modified') . ' <= ' . $db->quote($modified));
+        }
+
```
i had to read this far too many times to understand it. its confusing that the variable for modified_start and modified_end have the same name

laoneo replied:
Renamed

brianteeman replied:
thanks!!

laoneo committed:
clearer variables

alikon commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -492,6 +492,20 @@ protected function getListQuery()
+        $modifiedStartDateTime = $this->getState('filter.modified_start');
+
+        if (!empty($modifiedStartDateTime)) {
+            $query->where($db->quoteName('a.modified') . ' >= ' . $db->quote($modifiedStartDateTime));
```
please use  prepared statement

laoneo replied:
If you can say me how, then I will do. I tried it first and that didn't worked with STRING type.

alikon replied:
see https://github.com/Digital-Peak/joomla-cms/pull/31

alikon commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -492,6 +492,20 @@ protected function getListQuery()
+        // Filter by date after modified date.
+        $modifiedStartDateTime = $this->getState('filter.modified_start');
+
+        if (!empty($modifiedStartDateTime)) {
```
```php
        if (!empty($modifiedStartDateTime)) {
            $query->where($db->quoteName('a.modified') . ' >= :startDate')
                ->bind(':startDate', $modifiedStartDateTime);
        }

        // Filter by date before modified date.
        $modifiedEndDateTime = $this->getState('filter.modified_end');

        if (!empty($modifiedEndDateTime)) {
            $query->where($db->quoteName('a.modified') . ' <= :endDate')
                ->bind(':endDate', $modifiedEndDateTime);
        }
```

alikon committed:
prepared statement (#31)

alikon commented:
I have tested this item :white_check_mark: successfully on 24932536a8ec6036542c0ea46d7f50f4378ac09e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45142">issues.joomla.org/tracker/joomla-cms/45142</a>.</sub>

exlemor commented:
I have tested this item :white_check_mark: successfully on 24932536a8ec6036542c0ea46d7f50f4378ac09e<br /><br />Thanks to @alikon's and @richard67's help, I was able to successfully test this via Postman!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45142">issues.joomla.org/tracker/joomla-cms/45142</a>.</sub>

QuyTon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45142">issues.joomla.org/tracker/joomla-cms/45142</a>.</sub>

muhme commented:
I have tested this item :white_check_mark: successfully on 24932536a8ec6036542c0ea46d7f50f4378ac09e<br /><br />Tested with local Joomla 5.3-dev on macOS
* Before applying this PR:
  * Overall System Tests 538 passed
  * Tested API with HTTP Status 200 (unknown filters are ignored)
    * curl -w "\nHTTP Status: %{http_code}\n"  --header "X-Joomla-Token:${token}"  "${baseUrl}/api/index.php/v1/content/articles"
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-02-01&filter\[modified_end\]=2025-02-28"
* After applying this PR with Patch Tester
  * Overall System Tests 539 passed (including the new one)
  * Created three articles january, february and march and changed modified date in database to 15th and time to 12:00:00
  * checked
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-02-01&filter\[modified_end\]=2025-02-28" # Get only february article
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-01-01&filter\[modified_end\]=2025-04-01" # Get all three articles
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-02-01" # February and March article
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_end\]=2025-02-01" # Only January article
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-01-15%2012:00:00&filter\[modified_end\]=2025-03-15%2012:00:00" # All three articles
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-01-15%2012:00:00&filter\[modified_end\]=2025-03-15%2011:59:59" # January + February article
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-01-15%2012:00:00" # all three
    * curl -w "\nHTTP Status: %{http_code}\n" --header "X-Joomla-Token:${token}" "${baseUrl}/api/index.php/v1/content/articles?filter\[modified_start\]=2025-01-15%2012:00:01" # only februrary and march article<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45142">issues.joomla.org/tracker/joomla-cms/45142</a>.</sub>

brianteeman committed:
[5.3] Remove non-distributed files (#45152)

During the build process off a new release or update we remove various files that should not or do not need to be distributed

This PR extends the list of files to be removed from the libraries/vendor folder to include

.drone.yml
.drone.jsonnet

Examples can be seen in most of the libraries/vendor/joomla folders

Signed-off-by: BrianTeeman <brian@teeman.net>
Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

heelc29 committed:
[5.2] system test for multi-factor authentication (#44733)

Co-authored-by: Allon Moritz <allon.moritz@digital-peak.com>

laoneo committed:
[5.3] Major update idna converter (#45140)

laoneo committed:
[5.3] Major update dev dependencies (#45138)

laoneo committed:
[5.3] Update http package to version 3.1.0 to fix some deprecation notices (#45148)

laoneo committed:
Merge branch '5.3-dev' into api/content/modified

SniperSister committed:
Update TUF client to fix PHP 8.4 compat issues (#45166)

* Update php-tuf to version 5.3

* revert composer change

brianteeman committed:
[5.3] php warning in BackupcodesModel (#45170)

_this is an AI generated explanation_

The warning Only the first byte will be assigned to the string offset occurs because $temp1 is being treated as a string, and assigning a value to $temp1[$i] attempts to modify a specific character (byte) of the string. However, the value being assigned (random_int(0, 99999999)) is an integer, which is incompatible with this operation.

Root Cause:
In PHP, when you use array-like syntax (e.g., $temp1[$i]) on a variable that has not been explicitly declared as an array, PHP may treat it as a string. In this case, $temp1 is being treated as a string, and $temp1[$i] is interpreted as modifying a specific character of the string. Since random_int(0, 99999999) generates an integer, PHP raises a warning because you cannot assign an integer to a specific byte of a string.

Fix:
To resolve this issue, you need to explicitly declare $temp1 as an array before using it. This ensures that $temp1[$i] is treated as an array element rather than a string offset.

Explanation of the Fix:
Explicit Declaration:

$temp1 = []; ensures that $temp1 is treated as an array from the start.
This avoids PHP's automatic type juggling, which might treat $temp1 as a string if it is not explicitly declared.
Assigning Values:

$temp1[$i] = random_int(0, 99999999); now correctly assigns the random integer to the $ith index of the $temp1 array.

Summary:
This fix ensures that $temp1 is treated as an array, eliminating the warning and allowing the code to function as intended.

Signed-off-by: BrianTeeman <brian@teeman.net>

dgrammatiko committed:
Update client.es6.js (#45174)

richard67 committed:
Use Joomla\Filesystem\Helper as FilesystemHelper (#45139)

Don't use the deprecated CMS Filesystem but the framework package

laoneo committed:
Merge branch '5.3-dev' into api/content/modified

Bodge-IT committed:
Merge branch '6.0-dev' into api/content/modified

softforge commented:
Thank you @laoneo and the testers for all your work
